File: packages/examples/aws/test-client.py
#!/usr/bin/env python3
"""
Interactive test client for elizaOS AWS Lambda worker (Python)

Usage:
    python test-client.py --endpoint https://YOUR_API_ID.execute-api.us-east-1.amazonaws.com/prod/chat
    python test-client.py --endpoint http://localhost:3000/chat
"""

from __future__ import annotations

import argparse
import json
import sys
import time
from typing import TypedDict

import httpx


class ChatResponse(TypedDict):
    response: str
    conversationId: str
    timestamp: str


class ErrorResponse(TypedDict):
    error: str
    code: str


def parse_args() -> argparse.Namespace:
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        description="elizaOS AWS Lambda Test Client",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python test-client.py --endpoint https://abc123.execute-api.us-east-1.amazonaws.com/prod/chat
  python test-client.py --endpoint http://localhost:3000/chat
        """,
    )
    parser.add_argument(
        "--endpoint",
        required=True,
        help="API endpoint URL",
    )
    parser.add_argument(
        "--conversation",
        default=None,
        help="Resume existing conversation",
    )
    return parser.parse_args()


def send_message(
    client: httpx.Client,
    endpoint: str,
    message: str,
    conversation_id: str | None,
) -> ChatResponse:
    """Send a message to the API."""
    body: dict[str, str] = {"message": message}
    if conversation_id:
        body["conversationId"] = conversation_id

    response = client.post(
        endpoint,
        json=body,
        timeout=60.0,
    )

    data = response.json()

    if response.status_code != 200:
        error = data
        raise Exception(f"API Error ({error.get('code', 'UNKNOWN')}): {error.get('error', 'Unknown error')}")

    return data


def check_health(client: httpx.Client, base_endpoint: str) -> bool:
    """Check API health."""
    health_endpoint = base_endpoint.replace("/chat", "/health").rstrip("/")
    response = client.get(health_endpoint, timeout=10.0)

    if response.status_code == 200:
        data = response.json()
        print(f"✅ Connected to {data.get('runtime', 'unknown')} runtime (v{data.get('version', '?')})\n")
        return True
    return False


def main() -> None:
    """Main entry point."""
    args = parse_args()
    endpoint = args.endpoint
    conversation_id = args.conversation

    print("\n🤖 elizaOS AWS Lambda Test Client\n")
    print(f"📡 Endpoint: {endpoint}\n")

    with httpx.Client() as client:
        # Check health
        check_health(client, endpoint)

        print("💬 Chat with Eliza (type 'exit' to quit, 'new' for new conversation)\n")

        while True:
            try:
                user_input = input("You: ")
            except EOFError:
                break
            except KeyboardInterrupt:
                print("\n")
                break

            text = user_input.strip()

            if text.lower() in ("exit", "quit"):
                print("\n👋 Goodbye!")
                break

            if text.lower() == "new":
                conversation_id = None
                print("\n🔄 Starting new conversation...\n")
                continue

            if not text:
                continue

            print("Eliza: ", end="", flush=True)
            start = time.time()
            response = send_message(client, endpoint, text, conversation_id)
            duration = int((time.time() - start) * 1000)

            print(response["response"])
            print(f"\n  [{duration}ms | {response['conversationId'][:8]}...]\n")

            conversation_id = response["conversationId"]


if __name__ == "__main__":
    main()

File: packages/examples/aws/test-client.ts
#!/usr/bin/env npx ts-node
/**
 * Interactive test client for elizaOS AWS Lambda worker
 *
 * Usage:
 *   npx ts-node test-client.ts --endpoint https://YOUR_API_ID.execute-api.us-east-1.amazonaws.com/prod/chat
 *   npx ts-node test-client.ts --endpoint http://localhost:3000/chat
 */

import * as readline from "node:readline";

// Parse command line arguments
function parseArgs(): { endpoint: string; conversationId: string | null } {
  const args = process.argv.slice(2);
  let endpoint = "";
  let conversationId: string | null = null;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--endpoint" && args[i + 1]) {
      endpoint = args[i + 1];
      i++;
    } else if (args[i] === "--conversation" && args[i + 1]) {
      conversationId = args[i + 1];
      i++;
    } else if (args[i] === "--help" || args[i] === "-h") {
      console.log(`
elizaOS AWS Lambda Test Client

Usage:
  npx ts-node test-client.ts --endpoint <url> [options]

Options:
  --endpoint <url>      API endpoint URL (required)
  --conversation <id>   Resume existing conversation
  --help, -h            Show this help message

Examples:
  npx ts-node test-client.ts --endpoint https://abc123.execute-api.us-east-1.amazonaws.com/prod/chat
  npx ts-node test-client.ts --endpoint http://localhost:3000/chat
`);
      process.exit(0);
    }
  }

  if (!endpoint) {
    console.error("Error: --endpoint is required");
    console.error("Run with --help for usage information");
    process.exit(1);
  }

  return { endpoint, conversationId };
}

interface ChatResponse {
  response: string;
  conversationId: string;
  timestamp: string;
}

interface ErrorResponse {
  error: string;
  code: string;
}

async function sendMessage(
  endpoint: string,
  message: string,
  conversationId: string | null,
): Promise<ChatResponse> {
  const body: Record<string, string> = { message };
  if (conversationId) {
    body.conversationId = conversationId;
  }

  const response = await fetch(endpoint, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });

  const data = await response.json();

  if (!response.ok) {
    const error = data as ErrorResponse;
    throw new Error(`API Error (${error.code}): ${error.error}`);
  }

  return data as ChatResponse;
}

async function checkHealth(baseEndpoint: string): Promise<boolean> {
  // Derive health endpoint from chat endpoint
  const healthEndpoint = baseEndpoint.replace(/\/chat\/?$/, "/health");

  const response = await fetch(healthEndpoint, {
    method: "GET",
    headers: { "Content-Type": "application/json" },
  });

  if (response.ok) {
    const data = (await response.json()) as {
      runtime: string;
      version: string;
    };
    console.log(`✅ Connected to ${data.runtime} runtime (v${data.version})\n`);
    return true;
  }
  return false;
}

async function main(): Promise<void> {
  const { endpoint, conversationId: initialConversationId } = parseArgs();
  let conversationId = initialConversationId;

  console.log("\n🤖 elizaOS AWS Lambda Test Client\n");
  console.log(`📡 Endpoint: ${endpoint}\n`);

  // Check health
  await checkHealth(endpoint);

  // Create readline interface
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  console.log(
    "💬 Chat with Eliza (type 'exit' to quit, 'new' for new conversation)\n",
  );

  const prompt = (): void => {
    rl.question("You: ", async (input) => {
      const text = input.trim();

      if (text.toLowerCase() === "exit" || text.toLowerCase() === "quit") {
        console.log("\n👋 Goodbye!");
        rl.close();
        process.exit(0);
      }

      if (text.toLowerCase() === "new") {
        conversationId = null;
        console.log("\n🔄 Starting new conversation...\n");
        prompt();
        return;
      }

      if (!text) {
        prompt();
        return;
      }

      process.stdout.write("Eliza: ");
      const start = Date.now();
      const response = await sendMessage(endpoint, text, conversationId);
      const duration = Date.now() - start;

      console.log(response.response);
      console.log(
        `\n  [${duration}ms | ${response.conversationId.slice(0, 8)}...]\n`,
      );

      // Track conversation for continuity
      conversationId = response.conversationId;

      prompt();
    });
  };

  prompt();
}

main().catch((error) => {
  console.error("Fatal error:", error);
  process.exit(1);
});
